job-runner/chain/reaper.go

```diff
@@ -0,0 +1,630 @@
+	MakeRunning() JobReaper
+	MakeSuspended() JobReaper
+	MakeStopped() JobReaper
+}
+
+// Implements ReaperFactory, creating 3 types of reapers - for a
+// normally running chain, a stopped chain, or a suspended chain.
+type ChainReaperFactory struct {
+	Chain        *Chain
+	ChainRepo    Repo
+	Logger       *log.Entry
+	RMClient     rm.Client
+	RMCTries     int            // times to try sending info to RM
+	RMCRetryWait time.Duration  // time to wait between tries to send info to RM
+	DoneJobChan  chan proto.Job // chan jobs are reaped from
+	RunJobChan   chan proto.Job // (running reaper) chan jobs to run are sent to
+	RunnerRepo   runner.Repo    // (stopped + suspended reapers) repo of job runners
+}
+
+// Make a JobReaper for use on a running job chain.
+func (f *ChainReaperFactory) MakeRunning() JobReaper {
+	return &RunningChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			doneChan:          make(chan struct{}),
+			stopMux:           &sync.Mutex{},
+		},
+		runJobChan: f.RunJobChan,
+	}
+}
+
+// Make a JobReaper for use on a job chain being suspended.
+func (f *ChainReaperFactory) MakeSuspended() JobReaper {
+	return &SuspendedChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			doneChan:          make(chan struct{}),
+			stopMux:           &sync.Mutex{},
+		},
+		runnerRepo: f.RunnerRepo,
+	}
+}
+
+// Make a JobReaper for use on a job chain being stopped.
+func (f *ChainReaperFactory) MakeStopped() JobReaper {
+	return &StoppedChainReaper{
+		reaper: reaper{
+			chain:             f.Chain,
+			chainRepo:         f.ChainRepo,
+			rmc:               f.RMClient,
+			logger:            f.Logger,
+			finalizeTries:     f.RMCTries,
+			finalizeRetryWait: f.RMCRetryWait,
+			doneJobChan:       f.DoneJobChan,
+			stopChan:          make(chan struct{}),
+			doneChan:          make(chan struct{}),
+			stopMux:           &sync.Mutex{},
+		},
+		runnerRepo: f.RunnerRepo,
+	}
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for running chains.
+type RunningChainReaper struct {
+	reaper
+	runJobChan chan proto.Job // enqueue next jobs to run here
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if...
+// - chain is done: save final state + send to RM.
+// - job failed:    retry sequence if possible.
+// - job completed: prepared subsequent jobs and enqueue if runnable.
+func (r *RunningChainReaper) Run() {
+	defer close(r.doneChan)
```

Comment:
At first glance this looks dangerous to me because, in general, only senders should close channels, else you run the risk of the senders trying to send to a closed channel. Are we certain this is safe? Maybe a comment would help here